src/main/java/pl/agh/cs/io/WindowsListenerRunner.java

```diff
@@ -0,0 +1,44 @@
+package pl.agh.cs.io;
+
+import java.util.*;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+public abstract class WindowsListenerRunner {
+    private final static long POLLING_DELAY_MILLIS = 1000;
+    private static Consumer<OpenWindowsSnapshot> callback;
+    private static ScheduledExecutorService executorService = new ScheduledThreadPoolExecutor(1);
+
+    public static void run(Consumer<OpenWindowsSnapshot> callback) {
+        WindowsListenerRunner.callback = callback;
+        executorService.scheduleAtFixedRate(newGetOpenWindowsTask(), POLLING_DELAY_MILLIS, POLLING_DELAY_MILLIS, TimeUnit.MILLISECONDS);
+    }
+
+    private static Runnable newGetOpenWindowsTask() {
+        return () -> {
+            Window foregroundWindow = WindowsApiFacade.getForegroundWindow();
+            List<Window> windows = WindowsApiFacade.getOpenWindows();
+            Map<String, WindowsPerExe> exeWindows = toWindowsPerExeMap(windows);
+            WindowsPerExe foregroundWindowPerExe = exeWindows.remove(foregroundWindow.getExePath());
+            if (foregroundWindowPerExe == null)
+                foregroundWindowPerExe = WindowsPerExe.fromWindow(foregroundWindow);
+            OpenWindowsSnapshot snapshot = new OpenWindowsSnapshot(foregroundWindowPerExe, exeWindows);
+            callback.accept(snapshot);
+        };
+    }
+
+    private static Map<String, WindowsPerExe> toWindowsPerExeMap(List<Window> windows) {
```

Comment:
czy ta mapa nie trzyma podwójnie informacji o ścieżce?

Reply:
tak w ogóle to czytelniej by było zamknąć tę mapę i logikę do niej w jakiejś klasie(np WindowsPerExe). Dzięki temu inne klasy nie musiałyby znać jej struktury.

Reply:
> czy ta mapa nie trzyma podwójnie informacji o ścieżce?

Tak, ale scieżka potrzebna jest i w obiekcie i jako klucz, więc nie wiem czy jest sens inaczej.

> tak w ogóle to czytelniej by było zamknąć tę mapę i logikę do niej w jakiejś klasie(np WindowsPerExe). Dzięki temu inne klasy nie musiałyby znać jej struktury.

Tutaj to jest tylko metoda prywatna, a ta mapa leci to Snapshota